src/corelibs/U2Core/src/datatype/msa/MsaRow.cpp

```diff
@@ -806,9 +806,11 @@ QPair<ChromatogramData::TraceAndValue, ChromatogramData::TraceAndValue>
 
     auto peaksKeys = PEAKS.keys();
     const ChromatogramData& chromatogramData = *chromatogram;
+    ushort maxPeakValue = 0;
     for (auto peak : qAsConst(peaksKeys)) {
         const QVector<ushort>& chromatogramBaseCallVector = chromatogramData.*PEAKS.value(peak);
-        auto peakValue = chromatogramBaseCallVector[baseCall];
+        ushort peakValue = chromatogramBaseCallVector[baseCall];
+        maxPeakValue = qMax(maxPeakValue, peakValue);
```

Comment:
Это не максимальное значение пика а максимальное значение сигнала в этом месте.
Если это не пик - согласно имени этой функции эта точка должна игнорироваться.

Пример: 
Два пика в данном случае это 6% и 4%
![image](https://github.com/ugeneunipro/ugene/assets/17791039/6c6aa8ca-ce36-4b0a-875d-07d6d1486d54)

С другой стороны в комментарии к этой функции говорится:

`/** Returns pair of the first and the second most frequent chromatogram characters. */`
В этом случае вообще непонятно зачем анализировать форму графика.

Нужно исправить конфлике в описании и коде функции.

Reply:
> Это не максимальное значение пика а максимальное значение сигнала в этом месте.
> Если это не пик - согласно имени этой функции эта точка должна игнорироваться.
> Пример:
> Два пика в данном случае это 6% и 4%

Так функция сейчас так и работает. Она берет значения слева, справа и посередине и смотрит, чтобы значение посередине было выше значений слева и справа. Если это не так, то данный сигнал не рассматривается. Такой, базовый анализ. Более сложный анализ потребовал бы больше времени на реализацию, что было не очень рационально для задачи, выполняемой в рамках бесплатной поддержки. Однако, и он вполне неплохо работал на тех тестовых данных, что у меня были.

> С другой стороны в комментарии к этой функции говорится:
> /** Returns pair of the first and the second most frequent chromatogram characters. */
> В этом случае вообще непонятно зачем анализировать форму графика.

Да, описание не совсем точное, соглашусь, поправлю. Анализ формы графика нужен для того, чтобы мы отсекали плохо отсеквенированные символы (которые как раз таки и имеют подобный "волнистый" график), т.к. они неинформативны.

Reply:
Поводу комментария к функции - понятно.
Мне кажется здесь есть ошибка:
```
for (auto peak : qAsConst(peaksKeys)) {
        const QVector<ushort>& chromatogramBaseCallVector = chromatogramData.*PEAKS.value(peak);
        ushort peakValue = chromatogramBaseCallVector[baseCall];
        maxPeakValue = qMax(maxPeakValue, peakValue);
```

В данном случае ты еще не доказал что это пик:
```
    if (startValue <= peakValue && endValue <= peakValue) {
            peaks.append({peak, peakValue});
        }
```

но уже записал что maxPeakValue - является пиком.

Reply:
Наверное, я не правильно выразился. Мне стоило написать, что "Я добавил проверку, чтобы в паре хроматограм с высочайшими пиками всегда был тот, у которого значение **следа для данного символа** максимально". Переименовал переменную во избежании непоняток.

Reply:
Я прочитала обновленную документацию функции и пока еще вижу несоответствие.

Здесь объясняется что такое пики и что их вернется либо два, либо hasTwoPeaks будет false, правильно?

>> Returns pair of the first and the second peaks, wich chromatogram traces have the highest values of the current @position.
>>      * The word "peak" mean, that this trace should have the highest value in the middle of the base and lowest values on its edges.
>>      * If there are no two higest peaks, @hasTwoPeaks is set to false.  */

Тут ничего не говорится когда сигнал-не-пик в данной позиции выше одного или обоих пиков.

Эти детали надо добавить в описание функции, тогда можно будет ревьюить насколько реализация соответствует описанию.

Reply:
OK

Reply:
В новом описании:
>At least one of the peaks must have the maximum trail value for a given symbol - otherwise, hasTwoPeaks will be set to false.

Нужно заменить "символ" на "позицию". И далее возникает вопрос: почему только один пик должен быть выше а не оба? По сути 2 пика должны иметь симметричные качества, мы же заменяем один на другой и обратно в UI.
А здесь выходит несимметричное поведение: один из пиков может быть ниже не-пика а другой нет.

Пример:
A: 50% пик
C: 40% не пик
G: 6% пик
T: 4% не пик


В итоге алгоритм вернет A + G, но G будет ниже C

Предлагаю сделать обработку пиков симметричной - чтобы они оба были выше любого не-пик сигнала

Reply:
> Нужно заменить "символ" на "позицию"

OK

>  далее возникает вопрос: почему только один пик должен быть выше а не оба? 

Возникает обратный вопрос - а почему оба пика должны быть выше, а не только один. Ответ - потому что я решил, что так будет лучше. Правда в том, что данный функционал имеет только один достаточно четко характеризуемый сценарий - это когда у нас символ имеет два четких пика, занимающих первое и второе место по силе, а максимумы двух остальных следов хроматограммы ниже. В этом случае, все очень однозначно. А вот во всех остальных никакой четкой спецификации нет и поэтому я просто сделал, как мне показалось будет более правильным. Мне, формально, без разницы и я могу реализовать любой вариант.

Reply:
>Возникает обратный вопрос - а почему оба пика должны быть выше, а не только один. Ответ - потому что я решил, что так будет лучше. Правда в том, что данный функционал имеет только один достаточно четко характеризуемый сценарий - это когда у нас символ имеет два четких пика

Ну так и можно сделать именно этот понятный сценарий - 2 четких равнозначных пика. На остальное не возвращать пару вовсе. Это я и предложила выше - не выделять один пик от другого.

А откуда вообще взялась идея смотреть форму пика? Почему недостаточно видеть силу сигнала в данной точке?

Если мы к 2 самым сильным сигналам в данной точке добавляем еще threshold (что сила сигнала должна быть выше существенного уровня) то про форму вообще можно забыть и не гадать о производных: чем пик отличается от ямы или от склона. Такое решение было бы намного проще объяснить и не говорить что "об остальных случаях мы не додумали".

Reply:
> Ну так и можно сделать именно этот понятный сценарий - 2 четких равнозначных пика. На остальное не возвращать пару вовсе. Это я и предложила выше - не выделять один пик от другого.

Так сейчас так и делается. Но сейчас второй пик может оказаться ниже "ямки", а ты предлагаешь сделать так, чтобы два пика всегда были выше всего остального. Какой из вариантов правильный - я не знаю, т.к. просто не знаю, насколько важен пик, если он ниже "ямки". Меня любой вариант устраивает, могу без проблем сделать твой, если считаешь, что так будет лучше. 

> А откуда вообще взялась идея смотреть форму пика?

Отсутсвие пика означает, что про этот символ на этой позиции вообще ничего не известно из-а неточности секвенирования. Обычно  если на позиции находится ямка, пусть она и выше всех остальных пиков, все равно соответствующий ямке символ не стоит на данной позиции.

Reply:
> Ты предлагаешь сделать так, чтобы два пика всегда были выше всего остального...Меня любой вариант устраивает, могу без проблем сделать твой, если считаешь, что так будет лучше.

Если мы используем пики то это симметричное решение лучше, чем несимметричное

>Отсутсвие пика означает, что про этот символ на этой позиции вообще ничего не известно из-а неточности секвенирования.

Это не верно. В каждой точке у нас есть значение силы сигнала которую можно трактовать как вероятность нахождения символа. 

Пик это или не пик - это уже вторичная информация. Яма с вероятностью в 20% лучше чем пик с вероятностью в 5%. Поэтому наша изобретение искать пики скорее всего это ошибка выдаваемая за науку. На мой взгляд достаточно было бы работать не по производным (форме) а по первичным данным - силе сигнала + отсечке по  его минимальному значению.

Reply:
> Если мы используем пики то это симметричное решение лучше, чем несимметричное

Ок, сделано

> Все остальное

В свящи с тем, что я начинаю видеть кнопку "Load more..."  в середине данного треда, мне кажется, что он излишне разросся. Предлагаю тебе написать по пунктам все твои пожелания по изменению алгоритма и я их реализую.

Reply:
Лучший и самый понятный алгоритм на мой взгляд следующий:
Нужно выбрать 2 самых сильных сигнала в данной точке выше установленного порога.
Все

Ни форму сигналов, но другие производные в расчёт не брать, либо привести научную базу обосновывающую зачем мы это делаем.

Reply:
> Лучший и самый понятный алгоритм на мой взгляд следующий: Нужно выбрать 2 самых сильных сигнала в данной точке выше установленного порога. Все
> 
> Ни форму сигналов, но другие производные в расчёт не брать, либо привести научную базу обосновывающую зачем мы это делаем.

Что такое порог и как он работает? Какое у него значение и чем оно обосновано? Как я понимаю, именно этот порог является центральным элементом предлагаемого тобой алгоритма - без него появляется огромное количество ложноположительных срабатываний (в том числе и описанное в данной задаче никуда не девается). Эти срабатывания обусловлены тем, что в выравнивании очень много символов, не соответствующих следу, значение которого наивысшее на данном символе. Это происходит как раз таки у позиций с низким качеством секвенирования. Если же отсеять позиции, где присутствуют "ямки", то такие срабатывания сходят на нет.

Я тут подумал - возможно, самым верным решением будет вообще не учитывать только такие позиции, изначальные символы которых соответвуют самому высокому следу хроматограммы. В этом случае форма графика действительно становится не важна. Такой вариант устроит?

То есть еще раз поясню с картинкой:
![image](https://github.com/ugeneunipro/ugene/assets/26257527/448d1bb8-4476-40a2-9fb2-f38eb899d997)
На данной позиции у нас самый высокий след красный и соответствует Т, а символ у нас С. Значит, мы просто скипаем данную позицию.

![image](https://github.com/ugeneunipro/ugene/assets/26257527/a1dc8658-5c96-4c6f-96c1-55d2cc3027ad)

На данной позиции самый высокий след синий, что соответствует С и символ тоже С - рассматриваем позицию.

Reply:
Порог в алгоритме уже есть:

![image](https://github.com/ugeneunipro/ugene/assets/17791039/8a44b573-e8a9-426b-95fc-36d7fc1c4783)
Нового поняния вводить не нужно и по текущему порогу нужно отсекать маловерояные (слабые) значения сигнала

Примеры я не поняла. В последовательность полученной с хроматограмы всегда выбирается символ с наибольшим сигналом. В твоем первом варианте это красный сигнал (T) и альтернативный это синий (C) если подойдет по порогу.

Во втором графике главный сигнал это синий и альтернативный красный.

Reply:
По видимости, ты не до конца понимаешь, как работает данная функция. Порог, настраиваемый на вкладке, не связан с силой какого то конкретного отдельного сигнала. Значение порога указывает "необходимое значение второго по силе сигнала по отношению к первому по силу сигналу. Например, у нас есть следующие сигналы на какой-то абстрактной позиции:

- A - 100
- C - 75
- G - 50
- T - 25

Так же, на этой позиции стоит символ A. Ему соответствует сигнал со значением 100. То, что в рамках UGENE мы называем "альтернативная мутация" - это второй по силе сигнал (в нашем случае C). Он будет отображен, если его процентное соотношение к первому по силе сигналу больше или равно заданному порогу. То есть, а нашем случае это **75 / 100 * 100% = 75%**. То есть, если мы зададим порог 76% и выше, то на данной позиции все еще будет отображаться символ A. Если мы зададим 75% и ниже, то символ A заменится на символ C. 

> В последовательность полученной с хроматограмы всегда выбирается символ с наибольшим сигналом.

Нет, это не так. Если ты откроешь alignment.ugenedb, то ты там сможешь найти миллион контрпримеров. Один из них описан в данной задаче и как раз таки я пытаюсь придумать алгоритм, который бы хорошо работал для таких случаев и не вызывал ложных срабатываний.

Reply:
По поводу порога я не говорила о деталях реализации. Ты описал порог как формулу отношения силы сигналов: это вполне нормально и достаточно для фильтрации.

О контр примерах: можешь ли показать реальный abif файл где в качестве символа выбран не максимальный сигнал?

Предлагаемая мной идея проста: альтернативный сигнал это сигнал с максимальной силой удовлетворяющий требованиям порога.

Reply:
> О контр примерах: можешь ли показать реальный abif файл где в качестве символа выбран не максимальный сигнал?

Как я уже сказал, это, во-первых, символ, о котором идет речь в сценарии этой задачи. Конкретно - это тестовый репозиторий, файл "_common_data/sanger/sanger_10.ab1", позиция 980. График зеленый, что соответствует **A**, но стоит символ **G**:
![image](https://github.com/ugeneunipro/ugene/assets/26257527/b0e0f54f-a231-4f1e-9f5c-53346977d222)

А во-вторых, данный случай встречается очень часто. Если ты откроешь файл "data/samples/Sanger/alignment.ugenedb" и просмотришь риды беглым взглядом, то ты увидишь десятки (если не сотни) таких встречаемостей. Например, сравни самый высокий след хроматограммы и выставленный символ вот тут:
![image](https://github.com/ugeneunipro/ugene/assets/26257527/b22146c6-a1e5-4ee3-9e94-10b81b6ac413)
Таких мест так много, что перечислять их все я не буду.

> Предлагаемая мной идея проста: альтернативный сигнал это сигнал с максимальной силой удовлетворяющий требованиям порога.

Я это понял и, как я уже сказал, в общем случае такой алгоритм вызывает огромное количество ложноположительных срабатываний.

Reply:
Хорошо давай остановимся на симметричной фильтрации пиков о которой договорились выше.

Reply:
@yalgaer посмотри, пожалуйста.